x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/user/TransportHasPrivilegesAction.java

```diff
@@ -139,4 +196,17 @@ private static boolean testIndex(Automaton checkIndex, Automaton roleIndex) {
     private static boolean testPrivilege(Privilege checkPrivilege, Automaton roleAutomaton) {
         return Operations.subsetOf(checkPrivilege.getAutomaton(), roleAutomaton);
     }
+
+    private boolean testResourceMatch(String application, String checkResource, String checkPrivilegeName, Role userRole,
+                                      Collection<ApplicationPrivilege> privileges) {
+        final Set<String> nameSet = Collections.singleton(checkPrivilegeName);
+        final ApplicationPrivilege checkPrivilege = ApplicationPrivilege.get(application, nameSet, privileges);
```

Comment:
`ApplicationPrivilege.get` does some *O(n)* on the privileges arg, and `testResourceMatch` is called in a loop.
Do you think you can pull the grouping by privilege name out of the loop?

Reply:
This should be much easier once we have a separate class for privileges stored in the index, so I'm going to hold off on this until then.
There's definitely some major performance improvements we could do, but the API will be much simpler with a better object model (e.g. privileges will always have single name)